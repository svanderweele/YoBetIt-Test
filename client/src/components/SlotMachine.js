
import React from 'react';
import axios from 'axios';
import Table from 'react-bootstrap/Table';
import Swal from 'sweetalert2'
import Button from 'react-bootstrap/Button'
import SlotMachineItem from './SlotMachineItem';
import Card from 'react-bootstrap/Card';
import Collapse from 'react-bootstrap/Collapse';


class SlotMachine extends React.Component {


    constructor(props) {
        super(props)
        this.state = {
            totalSpins: 0,
            coins: 20,
            showScoreSheet: false,
            spinHistory: [
                {
                    results: ['cherry', 'banana', 'apple']
                },
                {
                    results: ['apple', 'cherry', 'apple']
                },
            ]
        };

        this.toggleScoreSheet = this.toggleScoreSheet.bind(this);
    }

    componentDidMount() {
    }

    toggleScoreSheet() {
        this.setState({ showScoreSheet: !this.state.showScoreSheet })
    }

    render() {

        return (
            <div>
                <div className="container">
                    <h1>Slot Machine</h1>
                    <Table striped bordered hover>
                        <thead>
                            <tr>
                                <th>Current Coins</th>
                                <th>Total Spins</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <td>{this.state.coins}</td>
                                <td>{this.state.totalSpins}</td>
                            </tr>
                        </tbody>
                    </Table>
                    <h3>History</h3>
                    <Table striped bordered hover>
                        <thead>
                            <tr>
                                <th>Result</th>
                                <th>Reward</th>
                            </tr>
                        </thead>
                        <tbody>
                            {this.state.spinHistory.map(spin => {
                                return <SlotMachineItem spin={spin} />
                            })}
                        </tbody>
                    </Table>
                    <Button variant="primary" className="mb-1" block>Spin</Button>
                    <Button variant="info" className="mb-1" block onClick={this.toggleScoreSheet}>Show Score Sheet</Button>
                    {/* <Card>
                        <Card>
                            <Card.Header>Coins {this.state.coins}</Card.Header>
                            <Card.Body>Total Spins {this.state.totalSpins}</Card.Body>
                        </Card>
                    </Card> */}
                    <Collapse in={this.state.showScoreSheet}>
                        <div className="mt-2">
                            <h3>Score Sheet</h3>
                            <Table striped bordered hover >
                                <thead>
                                    <tr>
                                        <th>Pattern</th>
                                        <th>Reward</th>
                                    </tr>
                                </thead>
                                <tbody>
                                    <tr>
                                        <td>2x apple</td>
                                        <td>10</td>
                                    </tr>
                                </tbody>
                            </Table>
                        </div>

                    </Collapse>

                </div>
            </div>
        );

    }
}

export default SlotMachine;