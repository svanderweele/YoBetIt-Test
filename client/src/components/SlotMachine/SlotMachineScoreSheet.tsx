import React from "react";
import { Table, Card } from "react-bootstrap";
import {
  SlotMachineRewardRequirement,
} from "../../models/SlotMachine";

interface SlotMachineScoreSheetProps {
  rewardRequirements: SlotMachineRewardRequirement[];
}

const SlotMachineScoreSheet = (props: SlotMachineScoreSheetProps) => {
  return (
    <Card>
      <Card.Body>
        <Card.Title>Slot Machine Score Sheet</Card.Title>
        <Table bordered>
          <thead>
            <tr>
              <th>Id</th>
              <th>Pattern</th>
              <th>Reward</th>
            </tr>
          </thead>
          <tbody>
            {props.rewardRequirements.map(
              (reward: SlotMachineRewardRequirement) => {
                return (
                  <tr key={reward.id}>
                    <td>{reward.id}</td>
                    <td>{reward.patternString}</td>
                    <td>{reward.reward}</td>
                  </tr>
                );
              }
            )}
          </tbody>
        </Table>
      </Card.Body>
    </Card>
  );
};

export default SlotMachineScoreSheet;
