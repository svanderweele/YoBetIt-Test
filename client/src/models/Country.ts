class Country {
  flag: string;
  name: string;
  activeCoronaCases: number;

  constructor(flag: string, name: string, activeCoronaCases: number) {
    this.flag = flag;
    this.name = name;
    this.activeCoronaCases = activeCoronaCases;
  }
}

class Response {
  message: string;
  success: boolean;
  constructor(success: boolean, message: string) {
    this.success = success;
    this.message = message;
  }
}

class GetCountriesResponse extends Response {
  countries: Country[];
  constructor(success: boolean, message: string, countries: Country[]) {
    super(success, message);
    this.countries = countries;
  }
}

export { Country, GetCountriesResponse };
