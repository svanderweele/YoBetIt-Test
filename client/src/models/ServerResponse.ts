class ServerResponse {
  success: boolean;
  message: string;
  data?: object;
  constructor(success: boolean, message: string, data?: object) {
    this.success = success;
    this.message = message;
    this.data = data;
  }
}

export default ServerResponse;
