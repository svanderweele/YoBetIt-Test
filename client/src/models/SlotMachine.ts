enum SlotMachinePatternTypeEnum {
  Apple = 0,
  Banana = 1,
  Cherry = 2,
  Lemon = 3,
}

class SlotMachineSpin {
  id: number;
  date: Date;
  result: string;
  userId: number;
  reward: number;

  constructor(
    id: number,
    date: Date,
    result: string,
    userId: number,
    reward: number
  ) {
    this.id = id;
    this.date = date;
    this.result = result;
    this.userId = userId;
    this.reward = reward;
  }
}

class SlotMachinePatternType {
  id: number;
  name: string;
  constructor(id: number, name: string) {
    this.id = id;
    this.name = name;
  }
}

class SlotMachineRewardRequirement {
  id:number;
  pattern: SlotMachinePatternTypeEnum[];
  patternString?: string;
  reward: number;
  constructor(id:number, pattern: SlotMachinePatternTypeEnum[], reward: number) {
    this.pattern = pattern;
    this.reward = reward;
    this.id = id;
  }
}

export { SlotMachinePatternType, SlotMachineSpin, SlotMachineRewardRequirement, SlotMachinePatternTypeEnum };
