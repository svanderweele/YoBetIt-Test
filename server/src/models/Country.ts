export class Country {
  id: number;
  flag: string;
  name: string;
  constructor(
    id: number,
    flag: string,
    name: string
  ) {
    this.id = id;
    this.flag = flag;
    this.name = name;
  }
}

export interface GetAllCountriesQuery {
  countryNames?: string;
  fields?: string;
  isExactNameMatch?: string;
}