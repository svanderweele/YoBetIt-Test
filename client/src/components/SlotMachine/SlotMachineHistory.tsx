import React from "react";
import { Table, Card } from "react-bootstrap";
import { SlotMachineSpin } from "../../models/SlotMachine";
import moment from 'moment';

interface SlotMachineHistoryProps {
  spins: SlotMachineSpin[];
}

const SlotMachineHistory = (props: SlotMachineHistoryProps) => {
  return (
    <Card>
      <Card.Body>
        <Card.Title>Slot Machine History</Card.Title>
        <Table bordered>
          <thead>
            <tr>
              <th>Result</th>
              <th>Reward</th>
              <th>Date</th>
            </tr>
          </thead>
          <tbody>
            {props.spins.map((spin: SlotMachineSpin) => {
              return (
                <tr key={spin.id}>
                  <td>{spin.result}</td>
                  <td>{spin.reward}</td>
                  <td>{moment(spin.date).format('MMMM Do YYYY, h:mm:ss a')}</td>
                </tr>
              );
            })}
          </tbody>
        </Table>
      </Card.Body>
    </Card>
  );
};

export default SlotMachineHistory;
