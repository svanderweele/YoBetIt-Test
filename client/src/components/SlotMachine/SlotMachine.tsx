import React from "react";
import { Card, Table, ButtonGroup, Button, Collapse } from "react-bootstrap";
import SlotMachineHistory from "./SlotMachineHistory";
import {
  SlotMachineSpin,
  SlotMachineRewardRequirement,
} from "../../models/SlotMachine";
import SlotMachineScoreSheet from "./SlotMachineScoreSheet";
import { User } from "../../models/User";
import toastr from "toastr";
import Swal, { SweetAlertResult } from "sweetalert2";
import ServerResponse from "../../models/ServerResponse";

const SlotMachine = () => {
  const [spinHistory, setSpinHistory] = React.useState<SlotMachineSpin[]>([
    new SlotMachineSpin(new Date(), "apple apple banana", 1, 20),
    new SlotMachineSpin(new Date(), "cherry cherry cherry", 1, 50),
  ]);

  const [isScoreSheetShown, setScoreSheetShown] = React.useState(false);
  const [userData, setUserData] = React.useState<User>({
    id: 0,
    name: "Simon van der Weele",
    money: 20,
  });

  const [rewardRequirements, setRewardRequirements] = React.useState<
    SlotMachineRewardRequirement[]
  >([]);

  const spin = () => {
    fetch(process.env.REACT_APP_HOST + "/api/slots/roll")
      .then((res) => res.json())
      .then((res: ServerResponse) => {
        if (res.success == false) {
          toastr.error(res.data.error);
        } else {
          let spin: SlotMachineSpin = res.data;
          if (spin.reward > 0) {
            toastr.success(`Won ${spin.reward.toString()}!`);
          } else {
            toastr.info(`No winnings`);
          }
        }

        getSpinHistory();
        getUserData();
      });
  };

  const getRewardRequirements = () => {
    fetch(process.env.REACT_APP_HOST + "/api/slots/score-sheet")
      .then((res) => res.json())
      .then((res: ServerResponse) => {
        setRewardRequirements(res.data);
      });
  };

  const getUserData = () => {
    fetch(process.env.REACT_APP_HOST + "/api/users/")
      .then((res) => res.json())
      .then((res: ServerResponse) => {
        setUserData(res.data[0]);
      });
  };

  const getSpinHistory = () => {
    fetch(process.env.REACT_APP_HOST + "/api/slots/spin-history")
      .then((res) => res.json())
      .then((res: ServerResponse) => {
        setSpinHistory(res.data);
      });
  };

  React.useEffect(() => {
    getSpinHistory();
    getUserData();
    getRewardRequirements();
  }, []);

  const reset = () => {
    Swal.fire({
      title: "Are you sure you want to reset?",
      text: "This will reset your history and money.",
      showCancelButton: true,
      background: "danger",
    }).then((result: SweetAlertResult) => {
      if (result.value) {
        fetch(`${process.env.REACT_APP_HOST}/api/slots/reset`)
          .then((res) => res.json())
          .then((res: ServerResponse) => {
            if (res.success) {
              toastr.success("Reset successful", "Successfully reset slots!");
              getSpinHistory();
              getUserData();
            } else {
              toastr.success("Reset error", res.data.error);
            }
          });
      }
    });
  };

  return (
    <Card>
      <Card.Body>
        <Card.Title>Slot Machine</Card.Title>
        <ButtonGroup aria-label="Basic example" className="mb-2">
          <Button variant="primary" onClick={spin}>
            Spin
          </Button>
          <Button
            variant="secondary"
            onClick={() => setScoreSheetShown(!isScoreSheetShown)}
            aria-controls="example-collapse-text"
            aria-expanded={isScoreSheetShown}
          >
            Show Score Sheet
          </Button>
          <Button variant="danger" onClick={reset}>
            Reset
          </Button>
        </ButtonGroup>
        <div>
          <h3>Current Money: {userData.money}</h3>
        </div>
        <SlotMachineHistory spins={spinHistory} />
        <hr />
        <Collapse in={isScoreSheetShown}>
          <div>
            <SlotMachineScoreSheet rewardRequirements={rewardRequirements} />
          </div>
        </Collapse>
      </Card.Body>
    </Card>
  );
};

export default SlotMachine;
