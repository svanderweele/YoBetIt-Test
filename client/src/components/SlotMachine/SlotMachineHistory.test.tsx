import React from "react";
import { shallow } from "enzyme";
import SlotMachineHistory from "./SlotMachineHistory";
import { SlotMachineSpin } from "../../models/SlotMachine";
describe("Slot Machine History", () => {
  it("should render correctly", () => {
    const spins = [
      new SlotMachineSpin(new Date(2020, 5, 28, 9, 15, 23), "Apple, Apple, Apple", 1, 20),
    ];

    spins.forEach((spin, index) => (spin.id = index));

    const component = shallow(<SlotMachineHistory spins={spins} />);
    expect(component).toMatchSnapshot();
  });
});
