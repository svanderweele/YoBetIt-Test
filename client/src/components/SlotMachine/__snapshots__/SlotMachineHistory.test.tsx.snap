// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot Machine History should render correctly 1`] = `
<Card
  body={false}
>
  <CardBody>
    <CardTitle>
      Slot Machine History
    </CardTitle>
    <ForwardRef
      bordered={true}
    >
      <thead>
        <tr>
          <th>
            Result
          </th>
          <th>
            Reward
          </th>
          <th>
            Date
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="0"
        >
          <td>
            Apple, Apple, Apple
          </td>
          <td>
            20
          </td>
          <td>
            June 28th 2020, 09:15:23
          </td>
        </tr>
      </tbody>
    </ForwardRef>
  </CardBody>
</Card>
`;
